TELCODOCS-1731 - Update low latency tuning docs

aireilly opened the pull request:
Refactoring Low latency tuning assembly into more manageable sections.

Version(s):
openshift-4.14+

Issue:
https://issues.redhat.com/browse/TELCODOCS-1731

Link to docs preview:
https://71797--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/low_latency_tuning/cnf-performing-platform-verification-latency-tests.html
https://71797--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html
https://71797--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.html
https://71797--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/low_latency_tuning/cnf-understanding-low-latency.html


QE review:
- [ ] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->


Note for Peer reviewers: 
* This PR might look ominous but it's really not. This is a high priority refactor PR that sets the ground for future content changes. What looks like new content here is in fact content that has been modularized out of parent assemblies with minimal rewrites.
* A lot of the files in the PR are just updated modules with "Module used in..." comments.
* SME comments on content will be actioned in https://issues.redhat.com/browse/TELCODOCS-1733

aireilly commented:
@MarSik Can you PTAL at this  Low latency tuning docs rework. This is based on the suggested changes from you and Ashley some time back. Note though it might looks like there are lots of changes in the PR, the content has not be changed or reworded. Content changes are minimal - mainly just moving stuff around and fixing xrefs..

MarSik commented:
@aireilly There are still some things to fix

1) Switch the order of Provisioning real-time and low latency workloads  and Tuning nodes for low latency with the performance profile . The user should first set up the nodes and then deploy the workloads.

2) Reducing NIC queues using the Node Tuning Operator belongs to the Tuning nodes section

3) There is a whole page about Creating a Performance profile that should be moved to the low latency chapter. Probably to the Tuning nodes sub-chapter.

MarSik commented:
4) Upgrading the performance profile to use device interrupt processing and Configuring a node for IRQ dynamic load balancing should be moved from the Workload sub-chapters to the Node tuning sub-chapter. Possibly under a common Interrupt handling topic (together with About support of IRQ affinity setting)

aireilly commented:
/retest

aireilly commented:
/retest

aireilly commented:
Hi @MarSik I think this is ready for another scan. Please start with the docs preview:
https://71797--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/low_latency_tuning/cnf-understanding-low-latency

Though it looks like loads of file have changed, they really haven't content should be 1:1 just moved around.

bartwensley commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,82 @@
+
+* For more information about recommended firmware configuration, see xref:../../edge_computing/ztp-vdu-validating-cluster-tuning.adoc#ztp-du-firmware-config-reference_vdu-config-ref[Recommended firmware configuration for vDU cluster hosts].
+
+[id="disabling-interrupt-processing-for-individual-pods_{context}"]
+=== Disabling interrupt processing for individual pods
+
+To achieve low latency for workloads, some (but not all) pods require the CPUs they are running on to not process device interrupts. A pod annotation, `irq-load-balancing.crio.io`, is used to define whether device interrupts are processed or not. When configured, CRI-O disables device interrupts when the pod is running.
+
+To disable interrupt processing for individual pods, ensure that `globallyDisableIrqLoadBalancing` is set to `false` in the performance profile. Then, in the pod specification, set the `irq-load-balancing.crio.io` pod annotation to `disable`.
+
+The following pod specification contains this annotation:
+[source,yaml]
+----
+apiVersion: performance.openshift.io/v2
+kind: Pod
+metadata:
+  annotations:
+      irq-load-balancing.crio.io: "disable"
+spec:
+    runtimeClassName: performance-<profile_name>
+...
+----
```
I admit I don't understand how the document structure works, but I see there is a separate module for the other annotations (e.g. modules/cnf-disabling-cpu-cfs-quota.adoc). Should this one be in a separate module as well? Or does the document end up looking the same in the end?

aireilly replied:
Doc looks the same in the end, but I've moved it as suggested for consistency.

bartwensley commented on scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc:
```diff
@@ -0,0 +1,98 @@
+
+Learn about the Performance Profile Creator (PPC) and how you can use it to create a performance profile.
+
+[NOTE]
+====
+Currently, disabling CPU load balancing is not supported by cgroup v2. As a result, you might not get the desired behavior from performance profiles if you have cgroup v2 enabled. Enabling cgroup v2 is not recommended if you are using performance profiles.
+====
```
As mentioned in another comment I think this has been addressed, but maybe this isn't the time to remove it?

MarSik replied:
Not in this PR, but we need to watch out for merge conflicts.

bartwensley reviewed (commented):
I added some comments and questions - apologies in advance if some of them don't make sense. :)

djuran commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,82 @@
+[id="disabling-interrupt-processing-for-individual-pods_{context}"]
+=== Disabling interrupt processing for individual pods
+
+To achieve low latency for workloads, some (but not all) pods require the CPUs they are running on to not process device interrupts. A pod annotation, `irq-load-balancing.crio.io`, is used to define whether device interrupts are processed or not. When configured, CRI-O disables device interrupts when the pod is running.
+
+To disable interrupt processing for individual pods, ensure that `globallyDisableIrqLoadBalancing` is set to `false` in the performance profile. Then, in the pod specification, set the `irq-load-balancing.crio.io` pod annotation to `disable`.
```
```suggestion
To achieve low latency for workloads, some (but not all) containers require that  the CPUs they are pinned to must not process device interrupts. A pod annotation, `irq-load-balancing.crio.io`, is used to define whether device interrupts are processed or not on the CPUs where the pinned containers of the Pod are running. When configured, CRI-O disables device interrupts where the pods containers are running.

To disable interrupt processing for cpus where containers belonging individual pods are pinned, ensure that `globallyDisableIrqLoadBalancing` is set to `false` in the performance profile. Then, in the pod specification, set the `irq-load-balancing.crio.io` pod annotation to `disable`.
```

shajmakh commented on modules/configuring_hyperthreading_for_a_cluster.adoc:
```diff
@@ -83,20 +84,20 @@ The reserved and isolated CPU pools must not overlap and together must span all
 ====
-Hyperthreading is enabled by default on most Intel processors. If you enable hyperthreading, all threads processed by a particular core must be isolated or processed on the same core.
+Hyper-Threading is enabled by default on most Intel processors. If you enable Hyper-Threading, all threads processed by a particular core must be isolated or processed on the same core.
 ====
```
Let's also note that when HT is enabled, all guaranteed pods must use multiples of the SMT level to avoid a noisy neighbor situation (in which the pod will fail with a relevant error). SMT is on when there are two threads per core; hence, the requested CPUs must be an even number.

MarSik replied:
Documented upstream - see full-pcpus-only here https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy-options

shajmakh reviewed (commented):
Thanks for this!
Two comments from initial review. I still need to validate the profiles configuration on a real cluster.

MarSik commented on modules/cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk.adoc:
```diff
@@ -0,0 +1,52 @@
+
+[NOTE]
+====
+Only disable CPU load balancing when the CPU manager static policy is enabled and for pods with guaranteed QoS that use whole CPUs. Otherwise, disabling CPU load balancing can affect the performance of other containers in the cluster.
```
It is accurate. Because all the conditions you talk about are satisfied even in the unlikely case where the cpu manager policy is set to disabled (= no pinning!).

MarSik commented:
Folks, the comments are great and I agree with pretty much everything you said. We have many places where text fixes are needed. However the goal here was to create a document structure that can be read top-down. We shouldn't be using this PR to fix every single bug we see in the docs. All of them should be fixed, but the logistics requires we do that in separate PRs when the changes are not related to the structural reorg. Otherwise this will never get finished in time..

MarSik commented on modules/cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk.adoc:
```diff
@@ -0,0 +1,52 @@
+
+:_mod-docs-content-type: PROCEDURE
+[id="cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk_{context}"]
+= Disabling CPU load balancing for DPDK
```
Yeah, this has been here for ages.. but as you asked below, this feature is mostly used with busy loop workloads. Because it only disables cpu load balancing = the scheduler tick.

MarSik commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,82 @@
+You use the performance profile configuration to make these changes.
+You can update the kernel to kernel-rt (real-time), reserve CPUs for cluster and operating system housekeeping duties, including pod infra containers, isolate CPUs for application containers to run the workloads, and disable unused CPUs to reduce power consumption.
+
+[WARNING]
+====
+When you tune nodes for low latency, the usage of execution probes in conjunction with applications that require guaranteed CPUs can cause latency spikes. Use other probes, such as a properly configured set of network probes, as an alternative.
+====
```
It is true for the pinned workloads. The exec probes do not have a specific cpu assigned and can interfere. So this warning should move to the creating workloads that work in real time section.

MarSik commented on modules/cnf-configuring-high-priority-workload-pods.adoc:
```diff
@@ -0,0 +1,55 @@
+
+:_mod-docs-content-type: PROCEDURE
+[id="cnf-configuring-high-priority-workload-pods_{context}"]
+= Configuring high priority workload pods
```
The annotations will work, there will just be no effect right? I would maybe use "Disabling power saving mode for high priority pods (or containers?)".

aireilly replied:
Reworded the title

MarSik commented on modules/cnf-configure_for_irq_dynamic_load_balancing.adoc:
```diff
@@ -36,152 +36,3 @@ spec:
 ====
 When you configure reserved and isolated CPUs, the infra containers in pods use the reserved CPUs and the application containers use the isolated CPUs.
```
It is almost correct, but confusing. The infra processes (not containers) are placed on reserved cpus. I think we should say something generic and link to the other section indeed.

aireilly replied:
Reworded slightly and and updated with link

MarSik commented on modules/cnf-configuring-workload-hints.adoc:
```diff
@@ -30,4 +31,53 @@
\ No newline at end of file
+====
+
+The following table describes how combinations of power consumption and real-time settings impact on latency.
+
+.Impact of combinations of power consumption and real-time settings on latency
+[cols="1,1,1,1",options="header"]
+|===
+|Performance Profile creator setting |Hint |Environment |Description
+
+|Default
+a|[source,terminal]
```
This PR is specifically trying to avoid content changes unrelated to the structure. #72761 should indeed be the place that addresses this, we just need to watch out for merge conflicts.

MarSik commented on modules/cnf-about-irq-affinity-setting.adoc:
```diff
@@ -60,4 +61,4 @@ $ find /proc/irq -name effective_affinity -printf "%p: " -exec cat {} \;
 
-Some drivers use `managed_irqs`, whose affinity is managed internally by the kernel and userspace cannot change the affinity. In some cases, these IRQs might be assigned to isolated CPUs. For more information about `managed_irqs`, see link:https://access.redhat.com/solutions/4819541[Affinity of managed interrupts cannot be changed even if they target isolated CPU].
\ No newline at end of file
+Some drivers use `managed_irqs`, whose affinity is managed internally by the kernel and userspace cannot change the affinity. In some cases, these IRQs might be assigned to isolated CPUs. For more information about `managed_irqs`, see link:https://access.redhat.com/solutions/4819541[Affinity of managed interrupts cannot be changed even if they target isolated CPU].
```
I wish I knew the list myself honestly (it changes between releases). Documenting it here would be too detailed and the KBase is probably better suited for that I think.

aireilly replied:
Added to https://issues.redhat.com/browse/TELCODOCS-1733

MarSik commented on modules/cnf-configuring-power-saving-for-nodes.adoc:
```diff
@@ -0,0 +1,77 @@
+      [sysfs]
+      /sys/devices/system/cpu/intel_pstate/max_perf_pct = <x> <1>
+----
+<1> The `max_perf_pct` controls the maximum frequency the `cpufreq` driver is allowed to set as a percentage of the maximum supported cpu frequency. This value applies to all CPUs. You can check the maximum supported frequency in `/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq`. As a starting point, you can use a percentage that caps all CPUs at the `All Cores Turbo` frequency. The `All Co ...
```
We should address this in a separate PR, but @bartwensley I did tell you this question would show up :D

aireilly replied:
Added this to https://issues.redhat.com/browse/TELCODOCS-1733

aireilly commented on modules/cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk.adoc:
```diff
@@ -0,0 +1,52 @@
+
+[NOTE]
+====
+Currently, disabling CPU load balancing is not supported with cgroup v2.
```
If all we need to do is remove this note I think we can do this here. WDYT @MarSik

aireilly replied:
I've removed this note. LMK if this doesn't work.

aireilly commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,184 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dynamic-irq-pod
```
What about `dynamic-realtime-pod`?

aireilly commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,184 @@
+        drop: [ALL]
+  nodeSelector:
+    node-role.kubernetes.io/worker-cnf: "" <4>
+  runtimeClassName: performance-dynamic-irq-profile <5>
```
What about `performance-dynamic-realtime-profile`?

aireilly commented on modules/cnf-node-tuning-operator-creating-workload-that-works-in-real-time.adoc:
```diff
@@ -0,0 +1,35 @@
+
+:_mod-docs-content-type: PROCEDURE
+[id="cnf-node-tuning-operator-creating-workload-that-works-in-real-time_{context}"]
+= Creating a workload that works in real-time
```
Let's just get rid. It's useless.

aireilly commented:
> @MarSik asked me to review this. Overall, I really like the new structure.
> 
> A few comments though:
> 
> 1. Notice we still have idle=poll in OVS-DPDK PP section.  Is this still valid?  Guidance I have gotten from Intel is as of Ice Lake (3rd gen xeon) this should no longer be necessary for adequate performance, and actually consumes a lot more energy as a result.  The Telco DU profile removes this setting unless workload hint highPowerConsumption is used, which I see we have addressed in those sections.  Thoughts?
> 2. Believe there is a templating error where we describe latency spikes and how they are generally not acceptable for {rds} workloads.  This looks to also be a bug in the live production docs (4.15 specifically).
> 3. Only other feedback I have is in the Performing latency tests for platform validation section: it might be non-trivial to a traditional user to understand that the kubeconfig consumed by podman should be in a deeper hierarchy.  I found in RHEL8/9 in my home directory this would cause SElinux to complain.  In the examples, we give the current directory, then /kubeconfig/kubeconfig.  I think it might be clearer specifying /kubeconfigs/kubeconfig-cluster and make that consistent throughout the procedure sections, running the various tools.
> 4. For the section titled "Troubleshooting errors with the cnf-tests container":  I would just add to open a Support Case here as a value add to the documentation, to get help with the container should a customer / partner need to.

Added these comments to https://issues.redhat.com/browse/TELCODOCS-1733

aireilly commented on modules/cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk.adoc:
```diff
@@ -0,0 +1,52 @@
+[id="cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk_{context}"]
+= Disabling CPU load balancing for DPDK
+
+Functionality to disable or enable CPU load balancing is implemented on the CRI-O level. The code under the CRI-O disables or enables CPU load balancing only when the following requirements are met.
```
Added these comments to https://issues.redhat.com/browse/TELCODOCS-1733

MarSik commented:
/lgtm with follow-ups per the above discussions handled in further PRs (this PR is supposed to handle the structure and move chapters to thematically grouped sections).

aireilly commented:
/label peer-review-needed

dfitzmau commented:
/remove-label peer-review-needed

/label peer-review-in-progress

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc:
```diff
@@ -0,0 +1,98 @@
+[role="_additional-resources"]
+.Additional resources
+
+* xref:../../scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc#cnf-configuring-high-priority-workload-pods_cnf-provisioning-low-latency[Configuring high priority workload pods]
```
This points to a section called "Disabling power saving mode for high priority pods"? Link title should match target.

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc:
```diff
@@ -0,0 +1,98 @@
+
+* xref:../../scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus_cnf-low-latency-perf-profile[Managing device interrupt processing for guaranteed pod isolated CPUs]
+
+include::modules/cnf-cpu-infra-container.adoc[leveloffset=+1]
```
More of a nice-to-have, but adding comments before each include that mentions the module title would be helpful as file names area bit more taxing to match up on the preview :).

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc:
```diff
@@ -0,0 +1,98 @@
+
+include::modules/cnf-cpu-infra-container.adoc[leveloffset=+1]
+
+include::modules/configuring_hyperthreading_for_a_cluster.adoc[leveloffset=+1]
```
underscores > hyphens ?

dfitzmau replied:
I do not see the module "hyyperthreading_for_low_latency_applications" listed in this assembly, but it should?

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,57 @@
+
+{product-title} provides the Node Tuning Operator to implement automatic tuning to achieve low latency performance and consistent response time for {product-title} applications.
+You use the performance profile configuration to make these changes.
+You can update the kernel to kernel-rt (real-time), reserve CPUs for cluster and operating system housekeeping duties, including pod infra containers, isolate CPUs for application containers to run the workloads, and disable unused CPUs to reduce power consumption.
```
Checking that "kernel-rt" does not need to be surrounded byt backticks?

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,57 @@
+
+[NOTE]
+====
+When writing your applications, follow the general recommendations described in link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/8/html-single/tuning_guide/index#chap-Application_Tuning_and_Deployment[Application tuning and deployment].
```
Link renders to a 404 page.

![Uploading Screenshot from 2024-05-02 12-44-28.png…]()

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,57 @@
+
+* xref:../../nodes/scheduling/nodes-scheduler-node-selectors.adoc#nodes-pods-node-selectors[Placing pods on specific nodes using node selectors]
+
+* link:https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node[Assigning pods to nodes]
```
```suggestion
* link:https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node[Assigning pods to nodes (Kubernetes documentation)]
```

Meets [this guideline](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#lead-in-link-sentence-style-for-the-link-attribute) and let's a customer know that they'll be visiting an external site in advance.

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc:
```diff
@@ -0,0 +1,57 @@
+[role="_additional-resources"]
+.Additional resources
+
+* For more information about recommended firmware configuration, see xref:../../edge_computing/ztp-vdu-validating-cluster-tuning.adoc#ztp-du-firmware-config-reference_vdu-config-ref[Recommended firmware configuration for vDU cluster hosts].
```
```suggestion
* xref:../../edge_computing/ztp-vdu-validating-cluster-tuning.adoc#ztp-du-firmware-config-reference_vdu-config-ref[Recommended firmware configuration for vDU cluster hosts].
```

bartwensley commented on modules/cnf-configure_for_irq_dynamic_load_balancing.adoc:
```diff
@@ -1,10 +1,10 @@
-// Module included in the following assemblies:
+  // Module included in the following assemblies:
```
nit: extra spaces?

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-debugging-low-latency-tuning-status.adoc:
```diff
@@ -0,0 +1,26 @@
+
+Use the `PerformanceProfile` custom resource (CR) status fields for reporting tuning status and debugging latency issues in the cluster node.
+
+include::modules/cnf-debugging-low-latency-cnf-tuning-status.adoc[leveloffset=+1]
```
cnf-debugging-low-latency-cnf-tuning-status-machineconfigpool is burried in a module but should in be listed in this assembly. Modules within modules might cause issues, but maybe I'm misunderstanding something here?

aireilly replied:
Will rework separately under https://issues.redhat.com/browse/TELCODOCS-1733

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-debugging-low-latency-tuning-status.adoc:
```diff
@@ -0,0 +1,26 @@
+
+include::modules/cnf-debugging-low-latency-cnf-tuning-status.adoc[leveloffset=+1]
+
+include::modules/cnf-collecting-low-latency-tuning-debugging-data-for-red-hat-support.adoc[leveloffset=+1]
```
Only on heading per module so should `About the must-gather tool` and `Gathering low latency tuning data` sections be brought to the assembly level?

Peer-review checklist 

![Screenshot from 2024-05-02 12-53-17](https://github.com/openshift/openshift-docs/assets/57954076/3fa96781-ad6d-452d-916b-d4bd65a79e52)

aireilly replied:
This would be a significant rewrite so will add this to the content rework PR

dfitzmau commented on scalability_and_performance/low_latency_tuning/cnf-debugging-low-latency-tuning-status.adoc:
```diff
@@ -0,0 +1,26 @@
+
+* xref:../../scalability_and_performance/using-node-tuning-operator.adoc#using-node-tuning-operator[Using the Node Tuning Operator]
+
+* xref:../../scalability_and_performance/what-huge-pages-do-and-how-they-are-consumed-by-apps.adoc#configuring-huge-pages_huge-pages[Configuring huge pages]
```
Configuring huge pages at boot time

bartwensley commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+// * scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.adoc
+
+[id="cnf-scheduling-workload-onto-worker-with-real-time-capabilities_{context}"]
+= Scheduling a workload onto a worker with real-time capabilities
```
Might be better as:
Scheduling a low latency workload onto a worker with real-time capabilities

bartwensley commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+[id="cnf-scheduling-workload-onto-worker-with-real-time-capabilities_{context}"]
+= Scheduling a workload onto a worker with real-time capabilities
+
+You can schedule realtime workloads onto a worker node where a performance profile that configures real-time capabilities is applied.
```
I wouldn't say "realtime" workload - linux doesn't support true realtime workloads. I think it is better to call these "low latency" workloads. Comment applies everywhere (assuming other folks agree with me).

bartwensley commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dynamic-realtime-pod
```
I would prefer "low-latency-pod". I don't think "dynamic" adds any value.

bartwensley reviewed (commented):
Just some comments about naming.

dfitzmau commented on _topic_maps/_topic_map.yml:
```diff
@@ -2949,14 +2949,21 @@ Topics:
+  - Name: Debugging low latency tuning
+    File: cnf-debugging-low-latency-tuning-status
+  - Name: Performing latency tests for platform verification
+    File: cnf-performing-platform-verification-latency-tests
```
This assembly/module is harder to locate on this PR. I cannot see it's sections listed either. Not sure why this is.

aireilly replied:
This assembly is untouched in this PR.

aireilly commented on modules/cnf-understanding-low-latency.adoc:
```diff
@@ -1,14 +1,11 @@
-Simply put, latency determines how fast data (packets) moves from the sender to receiver and returns to the sender after processing by the receiver. Maintaining a network architecture with the lowest possible delay of latency speeds is key for meeting the network performance requirements of 5G. Compared to 4G technology, with an average latency of 50 ms, 5G is targeted to reach latency numbers of ...
+= About low latency
 
 Many of the deployed applications in the Telco space require low latency that can only tolerate zero packet loss. Tuning for zero packet loss helps mitigate the inherent issues that degrade network performance. For more information, see link:https://www.redhat.com/en/blog/tuning-zero-packet-loss-red-hat-openstack-platform-part-1[Tuning for Zero Packet Loss in {rh-openstack-first}].
```
Added to https://issues.redhat.com/browse/TELCODOCS-1733

aireilly commented on modules/cnf-understanding-low-latency.adoc:
```diff
@@ -43,4 +40,3 @@ Workload hints simplify the fine-tuning of performance to industry sector settin
 In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload expectations and better able to meet the demands of the workload. The cluster admin can now specify into which use case that workload falls. The Node Tuning Operator uses the `PerformanceProfile` to fine tune the performance settings for the ...
 
 The environment in which an application is operating influences its behavior. For a typical data center with no strict latency requirements, only minimal default tuning is needed that enables CPU partitioning for some high performance workload pods. For data centers and workloads where latency is a higher priority, measures are still taken to optimize power consumption. The most complicated cases ...
```
Added to https://issues.redhat.com/browse/TELCODOCS-1733

mburke5678 commented on modules/cnf-about-hyperthreading-for-low-latency-and-real-time-applications.adoc:
```diff
@@ -0,0 +1,16 @@
+[id="cnf-about-hyper-threading-for-low-latency-and-real-time-applications_{context}"]
+= About Hyper-Threading for low latency and real-time applications
+
+Hyper-Threading is an Intel processor technology that allows a physical CPU processor core to function as two logical cores, executing two independent threads simultaneously. Hyper-Threading allows for better system throughput for certain workload types where parallel processing is beneficial. The default {product-title} configuration expects Hyper-Threading to be enabled by default.
```
>  The default {product-title} configuration expects Hyper-Threading to be enabled by default.
Is this redundant? 

```suggestion
Hyper-Threading is an Intel processor technology that allows a physical CPU processor core to function as two logical cores, executing two independent threads simultaneously. Hyper-Threading allows for better system throughput for certain workload types where parallel processing is beneficial. The default {product-title} configuration expects Hyper-Threading to be enabled.
```

mburke5678 commented on modules/cnf-about-hyperthreading-for-low-latency-and-real-time-applications.adoc:
```diff
@@ -0,0 +1,16 @@
+
+Hyper-Threading is an Intel processor technology that allows a physical CPU processor core to function as two logical cores, executing two independent threads simultaneously. Hyper-Threading allows for better system throughput for certain workload types where parallel processing is beneficial. The default {product-title} configuration expects Hyper-Threading to be enabled by default.
+
+For telecommunications applications, it is important to design your application infrastructure to minimize latency as much as possible. Hyper-Threading can slow performance times and negatively affect throughput for compute intensive workloads that require low latency. Disabling Hyper-Threading ensures predictable performance and can decrease processing times for these workloads.
```
```suggestion
For telecommunications applications, it is important to design your application infrastructure to minimize latency as much as possible. Hyper-Threading can slow performance times and negatively affect throughput for compute-intensive workloads that require low latency. Disabling Hyper-Threading ensures predictable performance and can decrease processing times for these workloads.
```

mburke5678 commented on modules/cnf-configure_for_irq_dynamic_load_balancing.adoc:
```diff
@@ -34,154 +34,8 @@ spec:
 [NOTE]
 ====
-When you configure reserved and isolated CPUs, the infra containers in pods use the reserved CPUs and the application containers use the isolated CPUs.
+When you configure reserved and isolated CPUs, operating system processes, kernel processes and systemd services run on reserved CPUs.
```
```suggestion
When you configure reserved and isolated CPUs, operating system processes, kernel processes, and systemd services run on reserved CPUs.
```

mburke5678 commented on modules/cnf-configure_for_irq_dynamic_load_balancing.adoc:
```diff
@@ -34,154 +34,8 @@ spec:
-When you configure reserved and isolated CPUs, the infra containers in pods use the reserved CPUs and the application containers use the isolated CPUs.
+When you configure reserved and isolated CPUs, operating system processes, kernel processes and systemd services run on reserved CPUs.
+Infrastructure pods run on any CPU except where the low latency workload is running.
+low latency workload pods run on exclusive CPUs from the isolated pool.
```
```suggestion
Low latency workload pods run on exclusive CPUs from the isolated pool.
```

mburke5678 commented on modules/cnf-configuring-hyperthreading-for-a-cluster.adoc:
```diff
@@ -83,20 +83,23 @@ The reserved and isolated CPU pools must not overlap and together must span all
+== Disabling Hyper-Threading for low latency applications
 
-When configuring clusters for low latency processing, consider whether you want to disable hyperthreading before you deploy the cluster. To disable hyperthreading, do the following:
+When configuring clusters for low latency processing, consider whether you want to disable Hyper-Threading before you deploy the cluster. To disable Hyper-Threading, do the following:
```
Per ISG: Use complete sentences, not phrases, to introduce lists.
```suggestion
When configuring clusters for low latency processing, consider whether you want to disable Hyper-Threading before you deploy the cluster. To disable Hyper-Threading, perform the following steps:
```

mburke5678 commented on modules/cnf-configuring-power-saving-for-nodes.adoc:
```diff
@@ -0,0 +1,77 @@
+
+.Prerequisites
+
+* You enabled C-states and Operating system controlled P-states in the BIOS
```
```suggestion
* You enabled C-states and operating system controlled P-states in the BIOS
```

mburke5678 commented on modules/cnf-configuring-power-saving-for-nodes.adoc:
```diff
@@ -0,0 +1,77 @@
+
+.Procedure
+
+. Generate a `PerformanceProfile` with `per-pod-power-management` set to `true`:
```
```suggestion
. Generate a `PerformanceProfile` with the `per-pod-power-management` argument set to `true`:
```

mburke5678 commented on modules/cnf-configuring-power-saving-for-nodes.adoc:
```diff
@@ -0,0 +1,77 @@
+--must-gather-dir-path /must-gather --power-consumption-mode=low-latency \ <1>
+--per-pod-power-management=true > my-performance-profile.yaml
+----
+<1> The `power-consumption-mode` must be `default` or `low-latency` when the `per-pod-power-management` is set to `true`.
```
```suggestion
<1> The `power-consumption-mode` argument must be `default` or `low-latency` when the `per-pod-power-management` argument is set to `true`.
```

mburke5678 commented on modules/cnf-configuring-power-saving-for-nodes.adoc:
```diff
@@ -0,0 +1,77 @@
+      [sysfs]
+      /sys/devices/system/cpu/intel_pstate/max_perf_pct = <x> <1>
+----
+<1> The `max_perf_pct` controls the maximum frequency the `cpufreq` driver is allowed to set as a percentage of the maximum supported cpu frequency. This value applies to all CPUs. You can check the maximum supported frequency in `/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq`. As a starting point, you can use a percentage that caps all CPUs at the `All Cores Turbo` frequency. The `All Co ...
```
```suggestion
<1> The `max_perf_pct` controls the maximum frequency that the `cpufreq` driver is allowed to set as a percentage of the maximum supported cpu frequency. This value applies to all CPUs. You can check the maximum supported frequency in `/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq`. As a starting point, you can use a percentage that caps all CPUs at the `All Cores Turbo` frequency. The `All Cores Turbo` frequency is the frequency that all cores will run at when the cores are all fully occupied.
```

mburke5678 commented on modules/cnf-configuring-workload-hints.adoc:
```diff
@@ -30,4 +31,53 @@
\ No newline at end of file
+====
+
+The following table describes how combinations of power consumption and real-time settings impact on latency.
```
```suggestion
The following table describes how combinations of power consumption and real-time settings impact latency.
```

mburke5678 commented on modules/cnf-disabling-cpu-cfs-quota.adoc:
```diff
@@ -0,0 +1,25 @@
+[id="cnf-disabling-cpu-cfs-quota_{context}"]
+= Disabling CPU CFS quota
+
+To eliminate CPU throttling for pinned pods, create a pod with the annotation `cpu-quota.crio.io: "disable"`. This annotation disables the CPU completely fair scheduler (CFS) quota when the pod runs. The following pod specification has this annotation:
```
```suggestion
To eliminate CPU throttling for pinned pods, create a pod with the `cpu-quota.crio.io: "disable"` annotation . This annotation disables the CPU completely fair scheduler (CFS) quota when the pod runs. The following pod specification has this annotation:
```

mburke5678 replied:
Or maybe?
```suggestion
To eliminate CPU throttling for pinned pods, create a pod with the `cpu-quota.crio.io: "disable"`  annotation. This annotation disables the CPU completely fair scheduler (CFS) quota when the pod runs. 
```
. Example pod specification with this annotation

mburke5678 commented on modules/cnf-node-tuning-operator-creating-pod-with-guaranteed-qos-class.adoc:
```diff
@@ -0,0 +1,71 @@
+
+:_mod-docs-content-type: PROCEDURE
+[id="cnf-node-tuning-operator-creating-pod-with-guaranteed-qos-class_{context}"]
+= Creating a pod with a QoS class of `Guaranteed`
```
Doc Guidelines: [Do not use backticks or other markup in assembly or module headings.](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings)

```suggestion
= Creating a pod with the Guranteed QoS class of Guaranteed
```
Or 
```suggestion
= Creating a pod with a Guaranteed QoS class
```

mburke5678 commented on modules/cnf-node-tuning-operator-creating-pod-with-guaranteed-qos-class.adoc:
```diff
@@ -0,0 +1,71 @@
++
+[NOTE]
+====
+If a container specifies its own memory limit, but does not specify a memory request, {product-title} automatically assigns a memory request that matches the limit. Similarly, if a container specifies its own CPU limit, but does not specify a CPU request, {product-title} automatically assigns a CPU request that matches the limit.
```
_specifies its own_  is a bit awkward to me. Is it even accurate?

```suggestion
If you specify a memory limit for a container, but do not specify a memory request, {product-title} automatically assigns a memory request that matches the limit. Similarly, you specify a  CPU limit for a container, but do not specify a CPU request, {product-title} automatically assigns a CPU request that matches the limit.
```

mburke5678 commented on modules/cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk.adoc:
```diff
@@ -0,0 +1,47 @@
+  runtimeClass: performance-manual
+----
+
+The Node Tuning Operator is responsible for the creation of the high-performance runtime handler config snippet under relevant nodes and for creation of the high-performance runtime class under the cluster. It will have the same content as default runtime handler except it enables the CPU load balancing configuration functionality.
```
```suggestion
The Node Tuning Operator is responsible for the creation of the high-performance runtime handler config snippet under relevant nodes and for creation of the high-performance runtime class under the cluster. It will have the same content as default runtime handler except that it enables the CPU load balancing configuration functionality.
```

mburke5678 commented on modules/cnf-running-the-performance-creator-profile.adoc:
```diff
@@ -82,10 +82,10 @@ Flags:
 [NOTE]
 ====
-Discovery mode inspects your cluster using the output from `must-gather`. The output produced includes information on:
+Discovery mode inspects your cluster by using the output from `must-gather`. The output produced includes information on:
```
Per ISG, sentences to introduce lists.
```suggestion
Discovery mode inspects your cluster by using the output from `must-gather`. The output produced includes information on the following conditions:
```

mburke5678 commented on modules/cnf-running-the-performance-creator-profile.adoc:
```diff
@@ -102,7 +102,7 @@ This command uses the performance profile creator as a new entry point to `podma
 The `-v` option can be the path to either:
```
```suggestion
The `-v` option can be the path to either of the following components:
```

mburke5678 commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+----
+<1> Disables the CPU completely fair scheduler (CFS) quota at the pod run time
+<2> Disables CPU load balancing
+<3> Opts the pod out of interrupt handling on the node
```
Periods
<1> Disables the CPU completely fair scheduler (CFS) quota at the pod run time.
<2> Disables CPU load balancing.
<3> Opts the pod out of interrupt handling on the node.

aireilly replied:
https://www.ibm.com/docs/en/ibm-style?topic=format-lists#punctuation-in-lists

Don't use periods when your list contains sentence fragments.

aireilly replied:
Actually no you're right. Full stops it is :sweat_smile:

mburke5678 commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+<4> The `nodeSelector` label must match the label that you specify in the `Node` CR.
+<5> `runtimeClassName` must match the name of the performance profile configured in the cluster.
+
+. Enter the pod `runtimeClassName` in the form performance-<profile_name>, where <profile_name> is the `name` from the `PerformanceProfile` YAML, in this example, `performance-dynamic-low-latency-profile`.
```
```suggestion
. Enter the pod `runtimeClassName` in the form performance-<profile_name>, where <profile_name> is the `name` from the `PerformanceProfile` YAML, in the previous example, the `name` is `performance-dynamic-low-latency-profile`.
```

mburke5678 commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+
+. Enter the pod `runtimeClassName` in the form performance-<profile_name>, where <profile_name> is the `name` from the `PerformanceProfile` YAML, in this example, `performance-dynamic-low-latency-profile`.
+
+. Ensure the pod is running correctly. Status should be `running`, and the correct cnf-worker node should be set:
```
> the correct cnf-worker node should be set

How do you determine which is the correct node and if it is set? I don't see anything in the output (maybe my ignorance).

aireilly replied:
The example isn't great. I've updated it slightly.

mburke5678 commented on modules/cnf-scheduling-workload-onto-worker-with-real-time-capabilities.adoc:
```diff
@@ -0,0 +1,187 @@
+----
+Cpus_allowed_list:  2-3
+----
+. Ensure the node configuration is applied correctly. Log in to the node to verify the configuration.
```
Add .Verification header. 


And, maybe
```suggestion
* Ensure the node configuration is applied correctly: 

.. Log in to the node to verify the configuration.
```
And turn all of the steps into substeps?

dfitzmau commented:
/remove-label peer-review-in-progress

/label peer-review-done

mburke5678 commented on modules/cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk.adoc:
```diff
@@ -0,0 +1,47 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  ...
```
For [snippets or sections of a file](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks), for YAML, use ('#…​')
```suggestion
#...
```

mburke5678 commented on modules/cnf-configuring-high-priority-workload-pods.adoc:
```diff
@@ -0,0 +1,55 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  ...
```
"For [snippets or sections of a file](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks)... for YAML, use ('#…​')" throughout this code block

```suggestion
# ...
```

mburke5678 commented on modules/cnf-disabling-cpu-cfs-quota.adoc:
```diff
@@ -0,0 +1,26 @@
+      cpu-quota.crio.io: "disable"
+spec:
+    runtimeClassName: performance-<profile_name>
+...
```
"For [snippets or sections of a file... for YAML](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks), use ('#…​')" throughout this code block

```suggestion
# ...
```

mburke5678 commented on modules/cnf-disabling-interrupt-processing-for-individual-pods.adoc:
```diff
@@ -0,0 +1,24 @@
+      irq-load-balancing.crio.io: "disable"
+spec:
+    runtimeClassName: performance-<profile_name>
+...
```
"For [snippets or sections of a file](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks)... for YAML, use ('#…​')" throughout this code block

```suggestion
# ...
```

mburke5678 commented:
@aireilly Found a few things. Awesome work here. Much easier to scan and use.

shajmakh commented:
LGTM from QE end considering this [comment](https://github.com/openshift/openshift-docs/pull/71797#issuecomment-2081988156).

aireilly commented:
/cherrypick enterprise-4.15

aireilly commented:
/cherrypick enterprise-4.16
